File: backend/app/api/routes/benchmark.py
"""Semgrep benchmark API (optional evaluation path — never part of the scan pipeline).

POST /api/benchmarks/semgrep   - run our engine + Semgrep on a controlled fixture
GET  /api/benchmarks          - read-only list of stored benchmark reports
GET  /api/benchmarks/{benchmark_id} - stored benchmark report

Errors: 404 (unknown fixture / unknown benchmark), 422 (invalid fixture name).
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, field_validator

from app.api.benchmark_models import BenchmarkList, BenchmarkSummary
from app.benchmark.ground_truth import get_ground_truth
from app.benchmark.models import BenchmarkReport
from app.benchmark.service import (
    BenchmarkService,
    InvalidFixtureNameError,
    UnknownFixtureError,
    get_report,
    list_reports,
)

router = APIRouter(prefix="/benchmarks", tags=["benchmark"])


class BenchmarkRequest(BaseModel):
    fixture: str

    @field_validator("fixture")
    @classmethod
    def _not_blank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("fixture must not be blank")
        return value.strip()


def _service() -> BenchmarkService:
    return BenchmarkService()


def _metric_f1(report: BenchmarkReport, tool: str) -> float | None:
    return next((m.f1 for m in report.metrics if m.tool == tool), None)


def _to_summary(report: BenchmarkReport) -> BenchmarkSummary:
    cases = get_ground_truth(report.fixture)
    vulnerable = len([c for c in cases if c.expected_vulnerable])
    return BenchmarkSummary(
        benchmark_id=report.benchmark_id,
        fixture=report.fixture,
        created_at=report.created_at,
        semgrep_available=report.semgrep_result.available,
        semgrep_error=report.semgrep_result.error,
        our_f1=_metric_f1(report, "our-sast"),
        semgrep_f1=_metric_f1(report, "semgrep"),
        ground_truth_cases=report.ground_truth_count,
        vulnerable_cases=vulnerable if cases else None,
        safe_cases=(len(cases) - vulnerable) if cases else None,
    )


@router.get("", response_model=BenchmarkList)
def list_benchmarks() -> BenchmarkList:
    reports = list_reports()
    return BenchmarkList(
        has_reports=bool(reports),
        reports=[_to_summary(r) for r in reports],
    )


@router.post("/semgrep", response_model=BenchmarkReport)
def run_benchmark(body: BenchmarkRequest) -> BenchmarkReport:
    try:
        return _service().run(body.fixture)
    except InvalidFixtureNameError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    except UnknownFixtureError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.get("/{benchmark_id}", response_model=BenchmarkReport)
def get_benchmark(benchmark_id: str) -> BenchmarkReport:
    report = get_report(benchmark_id)
    if report is None:
        raise HTTPException(
            status_code=404, detail=f"benchmark not found: {benchmark_id}"
        )
    return report

File: frontend/src/api/benchmark.ts
/* Typed API client for the Benchmark page. Mirrors the backend response
   models in app/benchmark/models.py and app/api/benchmark_models.py.
   The page is read-only; the only write operation is the intended backend
   execution endpoint POST /api/benchmarks/semgrep, which runs our engine
   and Semgrep against the controlled fixture server-side. The frontend
   never invokes Semgrep, shells or filesystem access itself. */

import { fetchJson } from "./dashboard";

export interface BenchmarkFinding {
  tool: string;
  vulnerability_type: string;
  file: string;
  line: number;
  function: string | null;
  message: string;
  fingerprint: string;
}

export interface BenchmarkResult {
  tool: string;
  available: boolean;
  findings: BenchmarkFinding[];
  duration_ms: number | null;
  error: string | null;
}

export interface BenchmarkMetrics {
  tool: string;
  true_positives: number;
  false_positives: number;
  false_negatives: number;
  precision: number | null; // 0-1 ratio, null when denominator is zero
  recall: number | null;
  f1: number | null;
  total_findings: number;
}

export interface BenchmarkComparison {
  shared_findings: BenchmarkFinding[];
  ours_only: BenchmarkFinding[];
  semgrep_only: BenchmarkFinding[];
  shared_vulnerability_types: string[];
  safe_cases_detected_incorrectly: string[];
}

export interface BenchmarkReport {
  benchmark_id: string;
  fixture: string;
  ground_truth_count: number;
  our_result: BenchmarkResult;
  semgrep_result: BenchmarkResult;
  metrics: BenchmarkMetrics[];
  comparison: BenchmarkComparison;
  created_at: string;
}

export interface BenchmarkSummary {
  benchmark_id: string;
  fixture: string;
  created_at: string;
  semgrep_available: boolean;
  semgrep_error: string | null;
  our_f1: number | null;
  semgrep_f1: number | null;
  ground_truth_cases: number;
  vulnerable_cases: number | null;
  safe_cases: number | null;
}

export interface BenchmarkList {
  has_reports: boolean;
  reports: BenchmarkSummary[];
}

export function getBenchmarkList(): Promise<BenchmarkList> {
  return fetchJson<BenchmarkList>("/api/benchmarks");
}

export function getBenchmarkReport(benchmarkId: string): Promise<BenchmarkReport> {
  return fetchJson<BenchmarkReport>(`/api/benchmarks/${benchmarkId}`);
}

export function runBenchmark(fixture: string): Promise<BenchmarkReport> {
  return fetch("/api/benchmarks/semgrep", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ fixture }),
  }).then((response) => {
    if (!response.ok) {
      throw new Error(`request failed: /api/benchmarks/semgrep (${response.status})`);
    }
    return response.json() as Promise<BenchmarkReport>;
  });
}
